Support for relative urls in batch payload #122

NilsEngelbach opened the pull request:
This pull request enables setting the `BatchPayloadUriOption` to support relative urls (#122).

```cs
var settings = new ODataClientSettings(httpClient, new Uri("odata/v1", UriKind.Relative))
{
    BatchPayloadUriOption = Microsoft.OData.BatchPayloadUriOption.RelativeUri
};
```

I will also provide unit tests soon.

watfordgnf commented on src/Simple.OData.Client.Core/ODataClientSettings.cs:
```diff
@@ -274,6 +274,14 @@ public IODataAdapterFactory AdapterFactory
         public bool ReadUntypedAsString { get; set; } = true;
 
+        /// <summary>
+        /// Gets or sets the BatchPayloadUriOption to use when building a batch request payload
```
```suggestion
        /// Gets or sets the BatchPayloadUriOption to use when building a batch request payload.
```

watfordgnf commented on src/Simple.OData.Client.Core/ODataClientSettings.cs:
```diff
@@ -274,6 +274,14 @@ public IODataAdapterFactory AdapterFactory
+        /// <summary>
+        /// Gets or sets the BatchPayloadUriOption to use when building a batch request payload
+        /// </summary>
+        /// <value>
+        /// <c>BatchPayloadUriOption.AbsoluteUri</c> (Default) to set the uri option for the batch payload
+        /// </value>
```
```suggestion
        /// <value>
        /// <c>AbsoluteUri</c> (Default) to use absolute URIs for the batch payload.
        /// </value>
        /// <value>
        /// <c>AbsoluteUriUsingHostHeader</c> to use absolute URIs from the Host header for the batch payload.
        /// </value>
        /// <value>
        /// <c>RelativeUri</c> to use relative URIs for the batch payload.
        /// </value>
```

watfordgnf reviewed (changes requested):
Few grammatical nits, however, we'll need a test or two for this.

Thank you for the submission!

NilsEngelbach commented:
@watfordgnf Thank you for the fast feedback.

I added 2 unit tests but i am not sure if they are added in the right place. I also adjusted the Mockdata "by hand" is there a way to generate the proper request/response mock somehow automatically?
Should this be tested with IntegrationTests instead?

I also recognized that the feature will only be available for OData V4, should this be added in the comment for the new setting?

NilsEngelbach reviewed (commented):
Applied all requested changes

NilsEngelbach commented:
@watfordgnf Can I do something more to complete this pull request?

watfordgnf commented:
@NilsEngelbach I haven't had an opportunity to pull this down locally and test it. I'm changing jobs in just over a week, so things are a bit busy :)

object commented:
@NilsEngelbach I ran tests for your PR and there were more than 300 unit test failures, Obviously something went wrong. Please make sure all tests pass.

Since the change can be tested using integration tests, can you write at least one such test for TripPinTests (part of Simple.OData.CLient.IntegrationTests)? And since this is OData 4 only change, perhaps it's worth to mention in XML comments for new ODataClientSettings.

NilsEngelbach commented:
@watfordgnf no worries and all the best for the new job, I was just surprised because your first response was so quick :)

@object I reverted my unit tests which were causing trouble with the other unit tests and added TripPin Integrationtests.

But it looks like the services does not support relative uris (the batch reader is missing the base uri) and returns the error:

```json
{
  "error":{
    "code": "InternalServerError",
    "message": "The relative URI 'Airlines' was specified in a batch operation, but a base URI was not specified for the batch writer or batch reader.",
    "innererror":{
      "message": "The relative URI 'Airlines' was specified in a batch operation, but a base URI was not specified for the batch writer or batch reader.",
      "type": "Microsoft.OData.Core.ODataException",
      "stacktrace": "at Microsoft.OData.Core.ODataBatchUtils.CreateOperationRequestUri(Uri uri, Uri baseUri, IODataUrlResolver urlResolver)
          at Microsoft.OData.Core.ODataBatchReader.ParseRequestLine(String requestLine, String& httpMethod, Uri& requestUri)
          at Microsoft.OData.Core.ODataBatchReader.CreateOperationRequestMessageImplementation()
          at Microsoft.OData.Core.ODataBatchReader.InterceptException[T](Func`1 action)
          at Microsoft.OData.Core.ODataBatchReader.CreateOperationRequestMessage()
          at Microsoft.Test.OData.Services.ODataWCFService.Handlers.BatchHandler.Process(IODataRequestMessage requestMessage, IODataResponseMessage responseMessage)
          at Microsoft.Test.OData.Services.ODataWCFService.Handlers.RequestHandler.Process(Stream requestStream)
          at Microsoft.Test.OData.Services.ODataWCFService.Handlers.RequestHandler.Process(Stream requestStream)
          at Microsoft.Test.OData.Services.ODataWCFService.Handlers.RootRequestHandler.Process(Stream requestStream)"
    }
  }
}
```

Who operates/maintains the TripPin Service?

I checked the request messages from the integration tests and they look correctly.
I also tried the new setting with my own OData service and it is working as expected...

object commented:
> @watfordgnf no worries and all the best for the new job, I was just surprised because your first response was so quick :)
> 
> @object I reverted my unit tests which were causing trouble with the other unit tests and added TripPin Integrationtests.
> 
> But it looks like the services does not support relative uris (the batch reader is missing the base uri) and returns the error:
> 
> ```json
> {
>   "error":{
>     "code": "InternalServerError",
>     "message": "The relative URI 'Airlines' was specified in a batch operation, but a base URI was not specified for the batch writer or batch reader.",
>     "innererror":{
>       "message": "The relative URI 'Airlines' was specified in a batch operation, but a base URI was not specified for the batch writer or batch reader.",
>       "type": "Microsoft.OData.Core.ODataException",
>       "stacktrace": "at Microsoft.OData.Core.ODataBatchUtils.CreateOperationRequestUri(Uri uri, Uri baseUri, IODataUrlResolver urlResolver)
>           at Microsoft.OData.Core.ODataBatchReader.ParseRequestLine(String requestLine, String& httpMethod, Uri& requestUri)
>           at Microsoft.OData.Core.ODataBatchReader.CreateOperationRequestMessageImplementation()
>           at Microsoft.OData.Core.ODataBatchReader.InterceptException[T](Func`1 action)
>           at Microsoft.OData.Core.ODataBatchReader.CreateOperationRequestMessage()
>           at Microsoft.Test.OData.Services.ODataWCFService.Handlers.BatchHandler.Process(IODataRequestMessage requestMessage, IODataResponseMessage responseMessage)
>           at Microsoft.Test.OData.Services.ODataWCFService.Handlers.RequestHandler.Process(Stream requestStream)
>           at Microsoft.Test.OData.Services.ODataWCFService.Handlers.RequestHandler.Process(Stream requestStream)
>           at Microsoft.Test.OData.Services.ODataWCFService.Handlers.RootRequestHandler.Process(Stream requestStream)"
>     }
>   }
> }
> ```
> 
> Who operates/maintains the TripPin Service?
> 
> I checked the request messages from the integration tests and they look correctly.
> I also tried the new setting with my own OData service and it is working as expected...

That's sad. TrinPinSerice is built and maintained by Microsoft. I suggest you keep the test but mark it as ignored, so it can be activated it they update the service implementation.

NilsEngelbach commented:
I hosted the TripPinService (https://github.com/OData/ODataSamples/tree/master/Scenarios/TripPin/src/webapi) locally and the Integrationtests are working with this local service.

I created an issue https://github.com/OData/ODataSamples/issues/140 and disabled the failing tests.

object commented:
Thank you for the PR. Merged and released as version 5.21.0
